hazelcast/src/main/java/com/hazelcast/map/impl/QueryResult.java

```diff
@@ -30,29 +32,36 @@
 
 public class QueryResult implements DataSerializable {
 
-    private final Collection<QueryResultEntry> result;
+    private final Collection<QueryResultEntry> result = new LinkedHashSet<QueryResultEntry>();
 
     private Collection<Integer> partitionIds;
 
+    private transient long resultLimit;
+    private transient long resultSize;
+
     public QueryResult() {
-        result = new LinkedHashSet<QueryResultEntry>();
+        this.resultLimit = Long.MAX_VALUE;
     }
 
-    @SuppressWarnings("unused")
-    public QueryResult(Collection<QueryResultEntry> queryableEntries) {
-        result = queryableEntries;
+    public QueryResult(long resultLimit) {
+        this.resultLimit = resultLimit;
     }
 
-    public Collection<Integer> getPartitionIds() {
-        return partitionIds;
+    public void addAll(Collection<QueryableEntry> queryableEntries) {
+        for (QueryableEntry entry : queryableEntries) {
+            if (++resultSize > resultLimit) {
+                throw new QueryResultSizeExceededException();
+            }
+            result.add(new QueryResultEntryImpl(entry.getKeyData(), entry.getIndexKey(), entry.getValueData()));
```

Comment:
I'm always careful with adding complex statements with deeper nesting. I personally prefer something like this:

```
QueryResultEntryImpl queryEntry = new QueryResultEntryImpl(
      entry.getKeyData(), entry.getIndexKey(), entry.getValueData())
result.add(queryEntry);
```

A bit more lines but less complex statements to deal with.